### PR TITLE
WIP Quobyte - Use Service Account as default

### PR DESCRIPTION
**Pull request by johscheuer**
**What this PR does / why we need it**:
This PR modifies the Quobyte Volume Plugin to use the `serviceaccount` name as `user` and `group`. This PR makes the plugin more secure/robust in an shared environment using service account names instead of letting the developer/user specify any user and group name they want. 

**Special notes for your reviewer**:
This will also works for the Dynamic Provisioning and create (if requested) a volume for the according service account as user/group.

FYI: @quolix @saad-ali @rootfs

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/35412)

### CODE REVIEW

**Comment by rootfs** [comment]
@liggitt

File: pkg/volume/quobyte/quobyte.go [hunk 1]
```diff
@@ -158,6 +158,23 @@ func (plugin *quobytePlugin) newMounterInternal(spec *volume.Spec, pod *api.Pod,
...
 	}
 
+	if pod != nil {
+		// We use the service account if it's not the default service account
+		if len(pod.Spec.ServiceAccountName) != 0 && pod.Spec.ServiceAccountName != "default" {
+			source.User = pod.Spec.ServiceAccountName
+			source.Group = pod.Spec.ServiceAccountName
+		} else {
+			// If the default service account is used or none is specified we use the specified user and group or default to root#nfsnobody
+			if len(source.User) == 0 {
```
Comment (rootfs) [inline pkg/volume/quobyte/quobyte.go]: defaults go to `pkg/api/v1/defaults.go`

File: pkg/volume/quobyte/quobyte.go [hunk 2]
```diff
@@ -158,6 +158,23 @@ func (plugin *quobytePlugin) newMounterInternal(spec *volume.Spec, pod *api.Pod,
 		return nil, err
 	}
 
+	if pod != nil {
+		// We use the service account if it's not the default service account
```
Comment (liggitt) [inline pkg/volume/quobyte/quobyte.go]: why special-case default?

File: pkg/volume/quobyte/quobyte.go [hunk 3]
```diff
@@ -158,6 +158,23 @@ func (plugin *quobytePlugin) newMounterInternal(spec *volume.Spec, pod *api.Pod,
...
+		if len(pod.Spec.ServiceAccountName) != 0 && pod.Spec.ServiceAccountName != "default" {
+			source.User = pod.Spec.ServiceAccountName
+			source.Group = pod.Spec.ServiceAccountName
```
Comment (liggitt) [inline pkg/volume/quobyte/quobyte.go]: are these supposed to map to named users on the node OS? that's not what service account is

Comment (liggitt) [inline pkg/volume/quobyte/quobyte.go]: > I couldn't find a way to get the Kubernetes user in the plugin

Kubernetes user info isn't available at this point (actually, I'm not really sure what you mean by that... the user that created the pod? last updated the pod? created the rc that resulted in this pod?)

> We know, actually it's enough to create a Quobyte Volume for the specified User/Group (service account name) for the mapping.

Can you clarify what source.User and source.Group are and how they are used? It looks like you are losing namespace info (so service account "foo" in ns1 and ns2 would appear identical)

File: pkg/volume/quobyte/quobyte.go (hunk 1 above)
Comment (johscheuer) [inline pkg/volume/quobyte/quobyte.go]: Ah thanks! Update is coming.

**Comment by johscheuer** [comment]
I added WIP to the PR name since there are some discussions on going.

File: pkg/volume/quobyte/quobyte.go (hunk 3 above)
Comment (liggitt) [inline pkg/volume/quobyte/quobyte.go]: > As Kubernetes user info I understand the "user" and "group" that are used to create the resource via the API (like with a static token file) and all "children" resources (pods) would inherit the user and group from the parent resource like a RC.

That is not how the system works today... if a user creates a RC, the system becomes responsible for fulfilling that RC's spec. If the user is removed from the system, but the RC remains, it should still be valid and pods should still be able to be created from it.

> These are just names that are used to map the access to the volume to, e.q. any access to a volume get mapped to the specified user and group (which are in the most cases unix user/groups).

Service account name is not connected in any way to fs users and groups today. I'd expect more discussion before committing to making that connection here. @pmorie, any feedback on how other volume plugins select owning uids/gids?

**Comment by johscheuer** [comment]
@liggitt 

> That is not how the system works today (and previous attempts to track creating users on objects have been less than successful - c.f. #13925)... if a user creates a RC, the system becomes responsible for fulfilling that RC's spec. If the user is removed from the system, but the RC remains, it should still be valid and pods should still be able to be created from it.

In our case this should work work fine, all we need is a Volume with an according user/group. I think you can image an Quobyte user similar to an RADOS user for Ceph (http://docs.ceph.com/docs/hammer/rados/operations/user-management). One challenge/issue we see is that if an Kubernetes user knows or guess an RADOS user he can access all resources the user has access too right e.q. when an user knows theres the RADOS user `admin` he can access all resources of the user `admin` by specifying `admin` as user in the `Volumes` part. (right @rootfs  ?)

So one possible solution would be to expose only PVC's to users and would prevent them to access "bare" volumes. AFAIK in OpenShift there's already a nice way to specify all these policies.

> Service account name is not connected in any way to fs users and groups today. I'd expect more discussion before committing to making that connection here. @pmorie, any feedback on how other volume plugins select owning uids/gids?

AFAIK at the moment there are no fine granular access policies for the most volume plugins (please correct me if I'm wrong) also I couldn't find anything inside the volume plugins (e.q. AWS, Azure, etc.). All I could find was the [security-context](this http://kubernetes.io/docs/user-guide/security-context)

Probably this would be a nice topic to talk about at the @kubernetes/sig-storage meeting tomorrow. 

I think this PR goes in a similar direction https://github.com/kubernetes/kubernetes/pull/35460 ?

File: pkg/volume/quobyte/quobyte.go [hunk 4]
```diff
@@ -158,6 +158,12 @@ func (plugin *quobytePlugin) newMounterInternal(spec *volume.Spec, pod *api.Pod,
 		return nil, err
 	}
 
+	// We use the service account if it's not the default service account
+	if pod != nil && len(pod.Spec.ServiceAccountName) != 0 && pod.Spec.ServiceAccountName != "default" {
```
Comment (liggitt) [inline pkg/volume/quobyte/quobyte.go]: the default service account shouldn't behave differently than other service accounts. I'm also not seeing how this adds protection in a shared env... the user has just as much control over the pod spec serviceAccountName as they do over the volume source user/group

**Comment by johscheuer** [comment]
@liggitt sorry for the long delay I was working on another project and only had some few spare time. Your points are all valid and I close this PR maybe in the future we will find a better way to add some more protection to volumes. For the current state the only (or easiest) way to add protection to volumes is to only allow `admins` to create volumes and all other "users" can only consume the volumes via `claims`.

Thanks for your time!

**Pull request closed without merging**